runtime/backend/options.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+#include <executorch/runtime/core/error.h>
+#include <executorch/runtime/core/span.h>
+#include <cstddef>
+#include <cstring>
+#include <variant>
+
+namespace executorch {
+namespace runtime {
+
+// Strongly-typed option key template
+template <typename T>
+struct OptionKey {
+  const char* key;
+  constexpr explicit OptionKey(const char* k) : key(k) {}
+};
+
+// Union replaced with std::variant
+using OptionValue = std::variant<bool, int, const char*>;
+
+struct BackendOption {
+  const char* key; // key is the name of the backend option, like num_threads,
+                   // enable_profiling, etc
+  OptionValue
+      value; // value is the value of the backend option, like 4, true, etc
+};
+
+template <size_t MaxCapacity>
+class BackendOptions {
+ public:
+  // Initialize with zero options
+  BackendOptions() : size_(0) {}
+
+  // Type-safe setters
+  template <typename T>
+  void set_option(OptionKey<T> key, T value) {
```

Comment:
rename this to set_option_impl and make it a private function. And add type safe public APIs to enforce key is always safe string... And value is also safe in the case of string.

```
  // String options - only accepts string literals/arrays
  template <size_t KeyN, size_t ValueN>
  Error set_option(
    const char (&key)[KeyN], 
    const char (&value)[ValueN]
  ) noexcept {
    return set_option_impl(key, static_cast<const char*>(value));
  }
  
  // Non-string options
  template <size_t N>
  Error set_option(const char (&key)[N], bool value) noexcept {
    return set_option_impl(key, value);
  }
  
  template <size_t N>
  Error set_option(const char (&key)[N], int value) noexcept {
    return set_option_impl(key, value);
  }
```